### DISCUSSION TITLE
general customisation guidance

### DISCUSSION THREAD

**Discussion by michaelssingh** [abougouffa/minemacs#128] [Q&A] [answered]
hey I have been meaning to give this a go and for the most part things look great, I have a few concerns, however, how do I control what is displayed when I hit SPC, what is the name of this package and where can I edit what appears, what is the recommended way to overwrite the built in module, in particular I want to add a few things to the clojure-mode that are missing

**Comment by abougouffa** [comment, marked as answer]
Hello @michaelssingh ,

You can take a look to [`skel/`](https://github.com/abougouffa/minemacs/tree/main/skel), I'm trying to keep it updated with configuration examples. The logic of user customization is explained briefly in the [README file](https://github.com/abougouffa/minemacs/tree/main?tab=readme-ov-file#customization). In a nutshell, it is recommended to configure packages in `with-eval-after-load` blocks, or by including a `use-package` block inside your `config.el` with your extra customization. The `config.el` is loaded after MinEmacs modules, so it can override what has been configured before. The order of loading configuration files is [described here](https://github.com/abougouffa/minemacs?tab=readme-ov-file#load-and-hooks-order).

Please note that MinEmacs uses `straight.el` to install packages, so it is recommended to use the same package manager and not mix it up with the builtin `package.el`. TLDR: use `:straight t` and not `:ensure t` in your `use-package` blocs.

**Comment by michaelssingh** [comment]
What about overwriting what you have here https://github.com/abougouffa/minemacs/blob/main/core/me-keybindings.el, I am trying to avoid forking this project.

**Comment by abougouffa** [comment]
If you want to overwrite the default keybindings you can do it using the `+map!` macro. This macro, among others (see `elisp/+keybinding.el`) are executed after `general` has been loaded and configured, as described here:

https://github.com/abougouffa/minemacs/blob/5c6739ef8dbd7c0f0d2f331d6a0b2a045563693b/core/me-keybindings.el#L225-L229

If you don't like how a particular package is configured and you want to disable its default configuration (provided by MinEmacs) and install it in your configuration file (`config.el`). You need to add it to the `minemacs-disabled-packages` in your `modules.el` file. The `skel/modules.el` contains some examples.

https://github.com/abougouffa/minemacs/blob/5c6739ef8dbd7c0f0d2f331d6a0b2a045563693b/skel/modules.el#L58-L73

I hope this answers your question!

**Comment by dorneanu** [comment]
Hi! I'm still not sure how to add additional packages and how to override some general behaviours. 
This is what I have in `~/.minemacs.d/local/config.el`: https://dpaste.org/9kYsP

For example I have: 

```lisp
(use-package embark
  :straight t
  :bind (("C-." . embark-act)
         :map embark-file-map
         ("V" . cust/vsplit-file-open)
         ("X" . cust/split-file-open))
  :config
  (define-key embark-url-map "X" #'dorneanu/embark-open-url-in-xwidget-webkit))

```

In general: How do I change the default keybindings (which by default use the leader `SPC`)? Sometimes I want to use `M-c` instead (for all the `cape` backends for example): 

```lisp
(use-package cape
  :straight t
  :demand t
  :hook ((prog-mode . dorneanu/set-basic-capf)
         (text-mode . dorneanu/set-basic-capf)
         (org-mode . dorneanu/set-basic-capf)
         ;; (go-ts-mode . dorneanu/set-go-capf)
         (lsp-completion-mode . dorneanu/set-lsp-capf))

  :general (:prefix "M-c"              ; I don't really use capitalize-word
                    "p" 'completion-at-point
                    "t" 'complete-tag   ; etags
                    "d" 'cape-dabbrev   ; basically `dabbrev-completion'
                    "f" 'cape-file
                    "k" 'cape-keyword
                    "s" 'cape-symbol
                    "a" 'cape-abbrev
                    "i" 'cape-ispell
                    "l" 'cape-line
                    "w" 'cape-dict
                    "\\" 'cape-tex
                    "_" 'cape-tex
                    "^" 'cape-tex
                    "&" 'cape-sgml
                    "r" 'cape-rfc1345)
  :custom
  (cape-dabbrev-min-length 1)
  :init
  ;; Add `completion-at-point-functions', used by `completion-at-point'.
  ;; (add-to-list 'completion-at-point-functions #'cape-history)
  ;; (add-to-list 'completion-at-point-functions #'cape-keyword)
  ;; (add-to-list 'completion-at-point-functions #'cape-dabbrev)
  ;; (add-to-list 'completion-at-point-functions #'cape-symbol)
  (add-to-list 'completion-at-point-functions #'cape-dabbrev)
  (add-to-list 'completion-at-point-functions #'cape-elisp-block)
  (add-to-list 'completion-at-point-functions #'cape-file)
)
```

The above example is from my current configuration (without minemacs). 
Any advice would be helpful. Thanks in advance.
